### DISCUSSION TITLE
How to disable purging in Tailwind 3?

### DISCUSSION THREAD

**Discussion by jaytoday** [tailwindlabs/tailwindcss#6557] [Help]
After upgrading to Tailwind 3, I would like to continue to be able to experiment locally in the browser with various CSS classes, but it appears that purging is removing all unused classes during compilation. It appears that `NODE_ENV=development` in `.env` is not making a difference. Are there ways of completely disabling purging so that all CSS is included in the compiled stylesheet?

**Comment by manhthang2504** [comment]
The only way is you include CDN

https://github.com/tailwindlabs/tailwindcss/discussions/6347

**Comment by leeshadle** [comment]
Creating a safelist using the pattern below seems to generate all the classes:

```
safelist: [
{
  pattern: /./ 
},
]
```

**Reply by Luis97lol** [reply]
That's great it worked for me. It would be even cooler if you could distinguish the environment variable so that you don't forget to purge. Thanks a lot

**Reply by vincerubinetti** [reply]
Thank you. This should really be documented. And better yet, there probably should be an `enabled` flag next to or within the new `content` option.

Also to link it to environment variable @Luis97lol , you can do something like this:

```
safelist: process.env.NODE_ENV === "development" ? [{ pattern: /.*/ }] : []
```

**Reply by Luis97lol** [reply]
Wow thats GREAT !!! Thank you very MUCH !!!

**Reply by leeshadle** [reply]
@Luis97lol and @vincerubinetti, I've been testing this slightly shorter pattern and it seems to be working great:

```
safelist: [
{
  pattern: /./ 
},
]
```

instead of:
```
safelist: [
{
  pattern: /.*/ 
},
]
```

**Reply by StevenGFX** [reply]
@vincerubinetti Thank you!!!

**Reply by sxshateri** [reply]
this doesn't work sometimes for me on Laravel. what is the point of such functionality at all? who needs to purge unused classes during dev? i just don't get it.

**Reply by adamwathan** [reply]
This video where we introduced the new engine explains all of the problems you run into if you have a large CSS file in dev: https://www.youtube.com/watch?v=3O_3X7InOw8

**Reply by rubalv96** [reply]
Thank you very much for the contribution. This works correctly, but in case we also want to have available the classes corresponding to dark mode this doesn't work as it purges these classes. For example, the ` .dark:text-green` class would not be recognised with this method. And, therefore, classes of type `dark:hover:text-green` would not be recognised either.

**Reply by zangab** [reply]
this does not include responsive utilities like `lg:my-4` unfortunately :/

**Reply by await-ovo** [reply]
Using this plugin should solve the problem of style changes not taking effect in Chrome Devtools:

https://www.npmjs.com/package/tailwindcss-webpack-plugin

**Reply by flyon** [reply]
yeah this solution does not work yet with any class with `:` in the name. So `md:grid-cols-2` or `dark:text-green` etc. Any solutions?

**Reply by StevenGFX** [reply]
> yeah this solution does not work yet with any class with `:` in the name. So `md:grid-cols-2` or `dark:text-green` etc. Any solutions?

I add in the variants: 

```
{
    pattern: /./,
    variants: ['sm', 'md', 'lg', 'xl', '2xl'],
},
```

**Reply by jswebschmiede** [reply]
for more infos: :)

[https://tailwindcss.com/docs/content-configuration#safelisting-classes](https://tailwindcss.com/docs/content-configuration#safelisting-classes)

**Reply by pel19072** [reply]
I was having this same issue for rendering grids with Tailwind CSS 3.3 with a React app created with react-create-app and this solved the problem. Here is the safelist I used:

`safelist: [
    {
      pattern: /grid-cols-(1|2|3|4)/,
      variants: ['sm'],
    },
  ]`

Thanks a lot for the help!

**Reply by davestewart** [reply]
FWIW adding `/./` to the safelist slowed our build process to a crawl then made Node run out of memory and crash after about 5 minutes on an M3 Macbook with 26GM RAM. Reproduced across various machines. YMMV.
